### DISCUSSION TITLE
Request to support RoundCorner in fvwm3

### DISCUSSION THREAD

**Discussion by qyangdu** [fvwmorg/fvwm3#946] [General]
## Describe the new feature

As the title, I used Round corner patch in fvwm2, I would like it to be supported in fvwm3 if possible. Thanks.

**Comment by somiaj** [comment]
@qyangdu If you can provide a copy of the patch, I'll take a look at it. Provided it can be implemented in such a way that doesn't alter normal functionality (making it an optional decor), we could see about adding it to fvwm3. Note a lot of those patches just weren't ever good enough to make it into the code, but I'll at least take a look at it.

**Comment by somiaj** [comment]
I grabbed the patch from archlinux, and tried to apply it to fvwm3. After fixing the conflicts, it doesn't build. I'm unsure if the issue is changes in fvwm3 or that this patch is built on top of other arch patches. It won't be an easy task to just apply the patch, so this feature may not be something that comes any time soon.

**Comment by qyangdu** [comment]
> I grabbed the patch from archlinux, and tried to apply it to fvwm3. After fixing the conflicts, it doesn't build. I'm unsure if the issue is changes in fvwm3 or that this patch is built on top of other arch patches. It won't be an easy task to just apply the patch, so this feature may not be something that comes any time soon.

Thanks for your update. The patch is for fvwm 2.6. There would be some work if there are a lot changes in fvwm3 comparing to fvwm2.

**Comment by qyangdu** [comment]
[0001-Add-round-corner-support.patch.gz](https://github.com/fvwmorg/fvwm3/files/10046576/0001-Add-round-corner-support.patch.gz)

Hi, I managed to make the patch compiled successfully and seems working normally in my notebook. Pls kindly help review and hope it could be supported in fvwm3 soon.

**Comment by somiaj** [comment]
This patch will still need a ton of work. The patch you have there modifies code which it should not. It is filled with parts that edit the tabs/spacing, which should be left alone. On top of that, as I mentioned above, the arch patches are built on top of each other, this patch also adds stuff for `BorderUnderTitle` and `FluxboxHandles`.

I also looked at the rounded corner effect, and though it does work, I still think it needs some work, it just chops off the hilight/shadow of 3D borders (instead of making them rounded), and modifying styles has unpredictable effect. Once you add the `RoundedCorners` style, if you remove it with `!RoundedCorners`, you are left with `FluxboxHandles`, and then you have to remove that, and even once you remove that, you are still left with artifacts of the rounded corners in the bottom of the window. Overall for general use this has a long ways to go.

**Comment by dumblob** [comment]
>Overall for general use this has a long ways to go.

There is (enough) time :wink:.

I am also looking forward to landing this after the behavior gets fixed.

**Comment by ONykyf** [comment]
Has there been any movement towards this?

**Comment by ONykyf** [comment]
I have rewritten the proposed above patch, with all related to fluxbox handles eliminated, and used Bresenham algorithm to make RoundedCorners of arbitrary radii, so now RoundedCorners style option can have up to 4 optional integer parameters (one for each corner, with reasonable substitutions if less are given), and RoundedCornersTop and RoundedCornersBottom up to 2 parameters. Patch is against fvwm3-1.0.9, but applies cleanly to the current snapshot. Just a proof of concept, but it works, please comment.
[fvwm3-1.0.9-roundedcorners.patch.gz](https://github.com/fvwmorg/fvwm3/files/14672635/fvwm3-1.0.9-roundedcorners.patch.gz)

**Comment by ONykyf** [comment]
XCalc is with RoundedCornersTop (default radii), XClock with RoundedCorners 10 18 25 0, taskbar at the top with RoundedCornersBottom 8 8.
![screenshot_rounded_corners](https://github.com/fvwmorg/fvwm3/assets/163148644/a44a9ecb-d96d-40be-9762-188d4e8d46f7)

Small ticks at the SW corner are because the radius is more than CornerLength, if the latter is increased, the artifacts disappear.
